src/ar_tx_replay_pool.erl

```diff
@@ -0,0 +1,290 @@
+	bhl = []
+}).
+
+-record(mempool, {
+	tx_id_set = sets:new()
+}).
+
+%% @doc Verify that a transaction against the given mempool, wallet list, recent
+%% weave txs, current block height, and current difficulty. The mempool is used
+%% to look for the same transaction there and to make sure the transaction
+%% does not reference another transaction from the mempool. The mempool is NOT
+%% used to verify shared resources like funds, wallet list references, and
+%% data size. Therefore, the function is suitable for on-edge verification
+%% where we want to accept potentially conflicting transactions to avoid
+%% consensus issues later.
+verify_tx(TX, Diff, Height, RecentTXs, MempoolTXs, WalletList) ->
+	WeaveState = create_state(RecentTXs),
+	TXIDSet = sets:from_list(
+		[MempoolTX#tx.id || MempoolTX <- MempoolTXs]
+	),
+	Mempool = #mempool {
+		tx_id_set = TXIDSet
+	},
+	verify_tx(
+		general_verification,
+		TX,
+		Diff,
+		Height,
+		WalletList,
+		WeaveState,
+		Mempool
+	).
+
+%% @doc Verify the transactions are valid for the block taken into account
+%% the given current difficulty and height, the previous blocks' wallet list,
+%% and recent weave transactions.
+verify_block_txs(TXs, Diff, Height, WalletList, RecentTXs) ->
+	WeaveState = create_state(RecentTXs),
+	{VerifiedTXs, _, _} = apply_txs(
+		TXs,
+		Diff,
+		Height,
+		WalletList,
+		WeaveState,
+		#mempool {}
+	),
+	case length(VerifiedTXs) of
+		L when L == length(TXs) ->
+			valid;
+		_ ->
+			invalid
+	end.
+
+%% @doc Pick a list of transactions from the mempool to mine on.
+%% Transactions have to be valid when applied on top of each other taken
+%% into account the current height and diff, recent weave transactions, and
+%% the wallet list. The total data size of chosen transactions does not
+%% exceed the block size limit. Before a valid subset of transactions is chosen,
+%% transactions are sorted from biggest to smallest and then from oldest
+%% block anchors to newest.
+pick_txs_to_mine(RecentTXs, Height, Diff, WalletList, TXs) ->
+	case ar_fork:height_1_8() of
+		H when Height >= H ->
+			pick_txs_to_mine_post_1_8(RecentTXs, Height, Diff, WalletList, TXs);
+		_ ->
+			pick_txs_to_mine_pre_1_8(Height, Diff, WalletList, TXs)
+	end.
+
+%% @doc Choose transactions to keep in the mempool after a new block is
+%% accepted. Transactions are verified independently from each other
+%% taking into account the given difficulty and height of the new block,
+%% the new recent weave transactions, and the new wallet list.
+pick_txs_to_keep_in_mempool(RecentTXs, TXs, Diff, Height, WalletList) ->
+	WeaveState = create_state(RecentTXs),
+	Mempool = #mempool{},
+	lists:filter(
+		fun(TX) ->
+			case verify_tx(
+				general_verification,
+				TX,
+				Diff,
+				Height,
+				WalletList,
+				WeaveState,
+				Mempool
+			) of
+				{valid, _, _} ->
+					true;
+				{invalid, _} ->
+					false
+			end
+		end,
+		TXs
+	).
+
+%% PRIVATE
+
+create_state(RecentTXs) ->
```

Comment:
Can we discuss this function together? I don't understand what `RecentTXs` is supposed to be...